src/puppetlabs/puppetdb/utils.clj

```diff
@@ -79,17 +79,17 @@
                      (.getPath (apply io/file export-root-dir file-suffix))
                      contents))
 
-(defn assoc-when
-  "Assoc to `m` only when `k` is not already present in `m`"
-  [m & kvs]
+(defmacro assoc-when
```

Comment:
I don't think assoc-when is a very good name for this functionality. I would imagine it only assoc-ing if the map is truthy, or something like that. This feels more like `lazy-assoc-defaults`, maybe?. Speaking of defaults, can we get most of the same behavior by leaning on the `:or` binding?

    (let [{:as m' :or {:a 1, :b 2}} m]
      m')

But this may not get us the shortcutting execution we want.